src/main/java/io/fabric8/jenkins/openshiftsync/BuildWatcher.java

```diff
@@ -345,40 +346,45 @@ public static synchronized boolean addEventToJenkinsJobRun(Build build)
         return false;
     }
 
-    public static synchronized void addBuildToNoBCList(Build build) {
+    private static void addBuildToNoBCList(Build build) {
         // should have been caught upstack, but just in case since public method
         if (!OpenShiftUtils.isPipelineStrategyBuild(build))
             return;
-        buildsWithNoBCList.add(build);
-    }
-
-    private static synchronized void removeBuildFromNoBCList(Build build) {
-        buildsWithNoBCList.remove(build);
+        try {
+          buildsWithNoBCList.add(build);
+        } catch (ConcurrentModificationException | IllegalArgumentException |
+          UnsupportedOperationException | NullPointerException e) {
+          logger.log(Level.WARNING,"Failed to add item " +
+            build.getMetadata().getName(), e);
+        }
     }
 
-    private static synchronized void clearNoBCList() {
-        buildsWithNoBCList.clear();
+    private static void removeBuildFromNoBCList(Build build) {
+          buildsWithNoBCList.remove(build);
     }
 
     // trigger any builds whose watch events arrived before the
     // corresponding build config watch events
-    public static synchronized void flushBuildsWithNoBCList() {
-        HashSet<Build> clone = (HashSet<Build>) buildsWithNoBCList.clone();
-        clearNoBCList();
-        for (Build build : clone) {
-            WorkflowJob job = getJobFromBuild(build);
-            if (job != null)
-                try {
-                    logger.info("triggering job run for previously skipped build "
-                            + build.getMetadata().getName());
-                    triggerJob(job, build);
-                } catch (IOException e) {
-                    logger.log(Level.WARNING, "flushCachedBuilds", e);
-                }
-            else
-                addBuildToNoBCList(build);
+    public static void flushBuildsWithNoBCList() {
+        for (Build build : buildsWithNoBCList) {
+          WorkflowJob job = getJobFromBuild(build);
+          if (job != null) {
+            try {
+              logger.info("triggering job run for previously skipped build "
+                + build.getMetadata().getName());
+              triggerJob(job, build);
+            } catch (IOException e) {
+              logger.log(Level.WARNING, "flushCachedBuilds", e);
+            }
+            try {
+              removeBuildFromNoBCList(build);
```

Comment:
no, you can't remove from a hash set within a for loop like this ... you can get a concurrent mod exception

You have to create the iterator, and then use iter.remove()

From my branch:

```
        Iterator<Build> iter = buildsWithNoBCList.iterator();
        while (iter.hasNext()) {
            Build build = iter.next();
            WorkflowJob job = getJobFromBuild(build);
            if (job != null) {
                try {
                    logger.info("triggering job run for previously skipped build "
                            + build.getMetadata().getName());
                    triggerJob(job, build);
                } catch (IOException e) {
                    logger.log(Level.WARNING, "flushCachedBuilds", e);
                }
                iter.remove();
            }
        }
```

Reply:
right, but this is the whole point I moved to ConcurrentHashSet.
AFAIK, Iterator needs syncronized, and it may kill the ConcurrentHashSet performance
blocking vs no-blocking

Actually by last run, I've compiled it with this ConcurrentHashSet and I see better results
so far I didn't see any concurrency problem with this change, what do you think?

Reply:
OK I pulled up the code for `org.eclipse.jetty.util.ConcurrentHashSet` and it is just a wrapper for `java.util.concurrent.ConcurrentHashMap` that implements the `Set` vs. `Map` semantics.

That, coupled with the protections around `ConcurrentModificationException` provided by `java.util.concurrent.ConcurrentHashMap`, I'm OK with this aspect of it.